// react-nav-bar/lib/lib/menuShapes.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.toggleShape = exports.springShape = undefined;

var _react = require('react');

/**
 * React shape object - used to animate menu open/close
 *
 * spring object spring={{ opened:100, closed:0, style: function(x) { return { height: `${x}%`} } }}
 */
var springShape = exports.springShape = _react.PropTypes.shape({
  style: _react.PropTypes.func.isRequired,
  opened: _react.PropTypes.number.isRequired,
  closed: _react.PropTypes.number.isRequired
});

/**
 * React shape object - used to control toggel button looks and visibility
 *
 *
 *
  var toggle = {
    default:'pied-piper',
    parent: {
      opened: 'bars',
      closed: 'pied-piper'
    },
    child: {
      opened: 'bus',
      closed: 'coffee'
    },
  };
 */
var toggleShape = exports.toggleShape = _react.PropTypes.shape({
  display: _react.PropTypes.bool,
  "default": _react.PropTypes.string,
  parent: _react.PropTypes.shape({
    opened: _react.PropTypes.string,
    closed: _react.PropTypes.string
  }),
  child: _react.PropTypes.shape({
    opened: _react.PropTypes.string,
    closed: _react.PropTypes.string
  })
});
